add control for showing a deck's track menu

ronso0 opened the pull request:
allows toggling a deck's track menu with controllers via `deck, show_track_menu` (read/write).
For now, `deck` is restricted to main decks (group name `[ChannelN]`)

This is an adaption of `[Library], show_track_menu`.
While the library menu requires a `WTrackTableView` instance, the deck control requires at least one `WTrackProperty` for each deck this control should be used with.

The tricky part was to allow only one open track menu per deck at a time. Check the comments for details.

For now, this is just a proposal. If there is interest to merge this I'm happy to receive reviews.

ronso0 committed:
librarycontrol: update window list

ronso0 committed:
WTrackMenu: add getDeckGroup()

ronso0 committed:
WTrackMenu: explicitly hide in closeEvent

ronso0 commented:
This works nicely now.
Mapped it for all decks to `Shift`+`Pfl`, and now I can delete obsolete track files and search related tracks from the controller with Pfl and the Trax turn/push encoder : )

daschuer commented:
This is still in draft state, is this still the case? 

Maybe the surprising behavior happens because of every WTrackProperty is connected to the new CO. 
Did you consider to only connect the first? 

I have not debugged it, but I can imagine that Mixxx tries to open many menus at once.

ronso0 commented:
This is a draft because it's still a proposal.

Please take a look at the code first. There are no surprises, it's just working.
* The first WTrackProperty that is parsed creates the push button and deals with the CO change requests.
* All other WTrackProperty widgets only connect the WTrackMenu's open/closed signal to a proxy.

This way there can be only track menu per deck:
* either opened by right-click on _any_ WTrackProperty
* or triggered by the CO, then it's the menu  of the first WTrackProperty (which doesn't need to be visbile btw, it should only be on the same side of the mixer as the other widgets)

ronso0 commented:
works great btw. I have now mapped it to Trax/Browse (press+hold) + deck Load button.
Since it is possible to open all deck (and library) track menus at once I made the script close all existing menus, then open the deck menu.
Pretty pleasant ux imo.

My main use cases are atill the Search Related menu and Remove from disk (5 button presses/turns to delete a track vs fiddling with the trackpad), and Properties which I can then easily navigate with the keyboard.

daschuer commented on src/widget/wtrackproperty.cpp:
```diff
@@ -123,3 +157,43 @@ void WTrackProperty::contextMenuEvent(QContextMenuEvent* event) {
+    // open/close events and from pushbutton changes.
+    // If the request matches the menu state we only set the control value accordingly.
+    // Otherwise we open/close the menu as requested. This will result in another
+    // change request originating from the menu , then we change the control value.
```
```suggestion
    // change request originating from the menu, then we change the control value.
```

daschuer commented on src/widget/wtrackproperty.cpp:
```diff
@@ -123,3 +157,43 @@ void WTrackProperty::contextMenuEvent(QContextMenuEvent* event) {
+        contextMenuEvent(&event);
+    } else {
+        pTrackMenu->close();
+    }
```
is a final  m_pShowTrackMenuControl->setAndConfirm(show); missing?

daschuer reviewed (changes requested):
Oh sorry that this is now conflicting. I have added some comments.

ronso0 commented:
I'm running into the same issue as with `stars_up/_down` (see #12591):
the ControlPushButtons are not destroyed on skin reload, i.e. they are not connected to the slot `WTrackProperty::slotShowTrackMenuChangeRequest`.

I noticed this with my personal branch but I didn't notice the warnings.

Therefore: back to the drawing board. I'll try moving the ControlPushButton creation to `BaseTrackPlayerImpl`.

ronso0 committed:
Track widgets: add control to open a deck's track menu

[ChannelN],show_track_menu

If show is requested all other track menus are closed.
Since this is not triggered by a mouse click there is no base point for the menu to be displayed.
The center of the first WTrackProperty widget is used instead, even if it's invisible.

ronso0 commented:
Overhauled and rebased to avoid the CO issue I encountered with the stars COs.

Works great now!

As usual, the last commit is for testing purposes only ; )
A MIDI mapping for testing with a virtual device is here https://gist.github.com/ronso0/78d506b0a0462dcdcbceb4f0453d72a8

daschuer reviewed (approved):
Works and locks good. Please remove the test commit than I will merge.

ronso0 commented:
Done.

Great, I can skip more and more commits when rebasing my personal branch onto main : )

daschuer commented:
Thank you.
